internal/sql/versioning/versioning.go

```diff
@@ -0,0 +1,107 @@
+// package versioning provides standard schema versioning for Kwil databases.
+package versioning
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kwilteam/kwil-db/common/sql"
+)
+
+var (
+	ErrTargetVersionTooLow = fmt.Errorf("target version is lower than current version")
+)
+
+// EnsureVersionTableExists ensures that the version table exists in the database.
+// If the table does not exist, it will be created, and the first version will be set to version specified.
+func ensureVersionTableExists(ctx context.Context, db sql.DB, schema string, targetVersion int64) error {
+	tx, err := db.BeginTx(ctx)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback(ctx)
+
+	// Create the schema if it does not exist
+	if _, err := tx.Execute(ctx, fmt.Sprintf(sqlCreateSchema, schema)); err != nil {
+		return err
+	}
+
+	// Create the version table if it does not exist
+	if _, err = tx.Execute(ctx, fmt.Sprintf(sqlVersionTable, schema)); err != nil {
+		return err
+	}
+
+	// Ensure that the version exists. If it does not, insert it with the target version.
+	if _, err = tx.Execute(ctx, fmt.Sprintf(sqlEnsureVersionExists, schema), targetVersion); err != nil {
+		return err
+	}
+
+	return tx.Commit(ctx)
+}
+
+// Upgrade upgrades the database to the specified version.
+// It will return an error if the database has already surpassed the specified version,
+// or if it is not possible to upgrade to the specified version.
+// All versions must be given as integers (e.g. 1, 2, 3, 4, 5, etc.), and are expected to be
+// sequential. A missing version will cause an error.
+// All upgrades will be transactional, and will be rolled back if an error occurs.
+// All versions should start at 0.
+// If the database is fresh, the schema will be initialized to the target version.
+// Raw initialization at the target version can be done by providing a function for versions -1.
+func Upgrade(ctx context.Context, db sql.DB, schema string, versions map[int64]UpgradeFunc, targetVersion int64) error {
+	tx, err := db.BeginTx(ctx)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback(ctx)
+
+	err = ensureVersionTableExists(ctx, tx, schema, targetVersion)
+	if err != nil {
+		return err
+	}
+
+	current, err := getCurrentVersion(ctx, tx, schema)
+	if err != nil {
+		return err
+	}
+
+	if current > targetVersion {
+		return fmt.Errorf(`%w: current version: %d, target version: %d`, ErrTargetVersionTooLow, current, targetVersion)
+	}
+
+	// Fresh DB, Raw initialization to the latest version
+	if current == targetVersion {
+		fn, ok := versions[-1]
+		if ok {
```

Comment:
Should this be `if current == -1` instead of `if current == targetVersion`?  If you're at the target version, you should do nothing, not init tables, right?

Reply:
Hmm, I think I now see why you did it this way, but I don't think it's ideal.  If `ensureVersionTableExists` ends up making the version table, we still need to do init tables.  So because you insert latest version into the fresh version table, you catch it this way.  Alternatively:
- create version table with -1/0 whatever indicates "fresh" and uninitialized
- OR have `ensureVersionTableExists` return a bool indicating if it created a new version table, and then decide to call init tables based on that.  Then simplify the logic down here.